sidecar on Kubernetes

nonsense opened the pull request:
This is our joint effort with @Stebalien on adding a K8sNetwork and basically adding support for a Kubernetes sidecar runner to Testground.

---

TODO:
- [x] ~decide whether we want to support specific IP allocation now, or we leave this feature for the future~ opened PR on weave's repo at https://github.com/weaveworks/weave/pull/3754
- [x] clean up
- [x] automate addition of ip route for the services CIDR, so that `redis-headless` and `redis-master` services are reachable from node's host networks.
- [x] clean up Dockerimage
- [x] change nonsens3 docker image repo to ipfs
- [x] fix routes code

---

MAYBE FOLLOW UP PR:

- [ ] confirm security privileges for initContainer that sets up route to k8s services CIDR
- [ ] check how to change default Weave network to 16.x.x.x - 32.x.x.x
- [ ] auto select a good network (ideally in the range 16. - 32. to match docker.)
- [ ] fix code which gets next available network - right now we have `10.33.10.0/24` hardcoded in `cluster_k8s.go`
- [ ] consider simplifying code that drops routes other than Redis in control network and the hardcoded podCIDR block - `100.96.0.0/11`. If we can't this needs to be documented with high visibility as it is not at all obvious!

---

Related:
1. https://github.com/coreos/flannel/issues/1243
2. https://github.com/weaveworks/weave/pull/3343

Stebalien commented on Dockerfile:
```diff
@@ -26,10 +28,11 @@ RUN cd / && CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o testground
 
 RUN mkdir -p /usr/local/bin
-COPY --from=0 /testground /usr/local/bin/testground
+RUN cp /testground /usr/local/bin/testground
```
?

nonsense replied:
Due to `iptables`. Will bring back the two-stage build Dockerfile, before merging, this was a shortcut.

Stebalien commented on Dockerfile:
```diff
@@ -26,10 +28,11 @@ RUN cd / && CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o testground
 #:::
 
-FROM busybox:1.31.0-glibc
+#FROM busybox:1.31.0-glibc
```
??

nonsense replied:
We needed `iptables` installed on the image, and this was the quickest way to get it. Not sure how you install it on `busybox` at the top of my head. Will fix this so that we have a small resulting image.

Stebalien replied:
That's fine. Adding iptables to busybox usually involves (a) installing it in debian then (b) copying it and hoping you have the right libc (that's what go-ipfs does).

Stebalien commented on infra/k8s/kops-weave/weave.yml:
```diff
@@ -163,6 +163,16 @@ items:
+          - name: iproute-add
+            image: busybox
+            securityContext:
+              privileged: true
```
We shouldn't need this, right?

nonsense replied:
I am not sure to be honest, I think I added it on purpose. Will double-check.

Ultimately I am not happy with this block being here at all, as it has nothing to do with `Weave`, but not sure where to place it. Ideally the issue on coreos/flannel will get some love, and we will understand better why this route is not there in the first place.

nonsense replied:
I've added it as a todo in the follow up (if I have time tomorrow) - we are just running `ip route` once, and then this process exits - I don't see any problem with not being specific with the `securityContext` and just giving root priveleges.

Stebalien commented on infra/k8s/sidecar.yaml:
```diff
@@ -2,28 +2,35 @@ apiVersion: apps/v1
       containers:
       - name: testground-sidecar
-        image: ipfs/testground:latest
+        image: nonsens3/testground:latest
```
TODO

nonsense replied:
Will change prior to merging. We are rebuilding the image on `master`, so during development I use my repo.

Stebalien commented on infra/k8s/sidecar.yaml:
```diff
@@ -2,28 +2,35 @@ apiVersion: apps/v1
         env:
         - name: REDIS_HOST
-          value: "redis-master"
+          value: "redis-headless"
```
Should we update our other runners?

nonsense replied:
This is actually redundant, because even though `redis-headless` returns a PodIP, in order to resolve it - the DNS is located on the services CIDR. So we are back at the problem with the missing route (which is now added through the `initContainer`).

Stebalien commented on infra/k8s/sidecar.yaml:
```diff
@@ -2,28 +2,35 @@ apiVersion: apps/v1
+        image: nonsens3/testground:latest
         command: ["testground"]
-        args: ["sidecar", "--runner", "docker"]
+        args: ["--vv", "sidecar", "--runner", "k8s"]
```
Do we want -vv to be the default?

Stebalien commented on pkg/runner/cluster_k8s.go:
```diff
@@ -96,8 +97,21 @@ func (*ClusterK8sRunner) Run(ctx context.Context, input *api.RunInput, ow io.Wri
 	}
 
+	var err error
+	_, runenv.TestSubnet, err = net.ParseCIDR("10.33.10.0/24")
```
TODO: auto select a good network (ideally in the range 16. - 32. to match docker.

nonsense replied:
That would involve changing the default Weave network. Will look into it.

nonsense replied:
I've added it as a TODO in the PR description for follow up.

Stebalien commented on pkg/sidecar/docker_instance.go:
```diff
@@ -307,7 +307,7 @@ func (dn *DockerNetwork) ConfigureNetwork(ctx context.Context, cfg *sync.Network
 				return err
 			}
-			delete(dn.activeLinks, netId)
+			delete(dn.activeLinks, cfg.Network)
```
Can you pull this into a separate PR so we can merge it first?

Stebalien commented on pkg/sidecar/sidecar.go:
```diff
@@ -7,11 +7,13 @@ import (
 
 var runners = map[string]func() (InstanceManager, error){
-	"docker": NewDockerManager,
+	//"docker": NewDockerManager,
```
?

nonsense replied:
Will revert. I had the wrong image installed, and before realising it was not clear whether the problem was the code or the wrong docker tag/image.

Stebalien commented on pkg/sidecar/k8s_instance.go:
```diff
@@ -0,0 +1,432 @@
+
+	return NewInstance(runenv, info.Config.Hostname, network)
+
+	// TODO(anteva): remove all routes but redis and the data subnet
```
Note: we're going to need to do this before the sidecar will be useful.

nonsense commented:
This has been tested with `network-pingpong` test, resulting in:

```
testground -vv run network/ping-pong \
    --builder=docker:go \
    --runner=cluster:k8s \
    --build-cfg bypass_cache=true \
    --build-cfg push_registry=true \
    --build-cfg registry_type=dockerhub \
    --run-cfg keep_service=true \
    --instances=2
```

```
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.10.0","Mask":"////AA=="}},"timestamp":15791939
17407652963,"msg":"I am 2"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
8310672062,"msg":"waiting until ready"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
8511295790,"msg":"writing my id"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
8511355509,"msg":"reading their id"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
8511367472,"msg":"returning their id"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
8511390826,"msg":"reading my id"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
8711973554,"msg":"done"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
8712039235,"msg":"ping RTT was 200.731959ms [200ms, 205ms]"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
9213567417,"msg":"waiting until ready"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
9213626067,"msg":"writing my id"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
9213641926,"msg":"reading their id"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
9234355791,"msg":"returning their id"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
9234387341,"msg":"reading my id"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
9234411425,"msg":"done"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
9234437160,"msg":"ping RTT was 20.803841ms [20ms, 30ms]"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"35f6af15-d835-4f47-8256-f693e1a27e2c","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.33.1.2","Mask":"////AA=="}},"timestamp":157919391
9485030356,"result":{"outcome":"ok"}}
```

nonsense commented on infra/k8s/kops-weave/weave.yml:
```diff
@@ -177,7 +191,8 @@ items:
                       fieldPath: spec.nodeName
-              image: 'docker.io/weaveworks/weave-kube:2.6.0'
+              #image: 'docker.io/weaveworks/weave-kube:2.6.0'
+              image: 'nonsens3/weave-kube'
```
@raulk @Stebalien note that this is using a forked version of the weave CNI plugin based on https://github.com/weaveworks/weave/pull/3754

Stebalien commented on pkg/sidecar/k8s_instance.go:
```diff
@@ -0,0 +1,386 @@
+
+		cniArgs := [][2]string{}                   // empty
+		capabilityArgs := map[string]interface{}{} // empty
+		ifName := "eth1"
```
Let's make constants for "eth1" and "eth2" (control network interface, data network interface).

Stebalien commented on pkg/sidecar/k8s_instance.go:
```diff
@@ -0,0 +1,386 @@
+	return networks
+}
+
+func (n *K8sNetwork) ListAvailable() []string {
```
We can drop this.

nonsense replied:
Not sure what you mean here - this is required by the interface.

Stebalien replied:
We can drop it from the interface as well. I think I pushed a patch to do that but it looks like it got clobbered.

ListActive is useful. I don't have any immediate uses for ListAvailable.

nonsense commented on infra/k8s/README-do-doks.md:
```diff
@@ -55,13 +55,13 @@ kubectl apply -f sidecar.yaml
 ## Run a Testground testplan
```
This README is going away - see next PR.

Stebalien replied:
Could we rename this to "README-deprecated.md" and add a comment?

nonsense replied:
Currently there are a few setups that we experimented with, with only one of them having a working sidecar:
1. AWS EKS (the Terraform playbooks) - one-click install for a k8s cluster on AWS EKS
2. Digital Ocean DOKS setup - manual install on Digital Ocean though the Web interface
3. AWS `kops` with 2 Flannel networks
4. AWS `kops` with Weave/Flannel networks

In the follow up PR I am removing everything but 4. because it is the only setup that has a working sidecar, and looks like we are going down that path.

---

So we can either keep various README-s for 1,2,3 in the codebase as `deprecated` (maybe in a `deprecated` folder), or we can remove them. I opted for the second (removing them in the next PR) because they really don't pass our requirements, and would like to keep our docs to the point.

I could keep them as you suggest as `README-deprecated.md` just wanted to share my thought process :)

Stebalien replied:
Got it. Sorry, I read "do-doks" as some kind of weird way to say "do the docs". We can probably leave it this way for now.

nonsense commented on pkg/sidecar/k8s_instance.go:
```diff
@@ -0,0 +1,386 @@
+
+	gw := redisRoute.Gw
+
+	podCidr := "100.96.0.0/11"
```
Moved this to `const` block, next to interface names `eth0` and `eth1`. There is probably a way to infer it, but this code is pretty much dependent on the networking setup of the cluster - same as the `eth0` and `eth1` names... I am not sure how much value there is to be smart and try to infer this block, as this code will most probably break if we change something on `weave` side or on Kubernetes side anyways.

nonsense commented:
@Stebalien I think I addressed all your comments. I have moved some of the tasks for follow up PR, because I think this one is already an improvement to `master`, even though it is not perfect.
